tables/TaQL/ExprNodeSetOpt.cc

```diff
@@ -0,0 +1,450 @@
+      if (value >= this->itsStarts[index]) {
+        return index;
+      }
+    }
+    return -1;
+  }
+  
+
+  template<typename T>
+  TableExprNodeSetOptContSetCO<T>::TableExprNodeSetOptContSetCO
+  (const TableExprNodeSet& orig,
+   const std::vector<T>& starts,
+   const std::vector<T>& ends)
+    : TableExprNodeSetOptContSet<T> (orig, starts, ends)
+  {}
+
+  template<typename T>
+  void TableExprNodeSetOptContSetCO<T>::show (ostream& os, uInt indent) const
+  {
+    TableExprNodeSetOptContSet<T>::show (os, indent);
+    os << " as TableExprNodeSetOptContSetCO" << endl;
+  }
+  
+  template<typename T>
+  Int64 TableExprNodeSetOptContSetCO<T>::find (T value) const
+  {
+    auto iter = std::upper_bound (this->itsEnds.begin(), this->itsEnds.end(), value);
+    if (iter != this->itsEnds.end()) {
+      size_t index = std::distance (this->itsEnds.begin(), iter);
+      if (value >= this->itsStarts[index]) {
+        return index;
+      }
+    }
+    return -1;
+  }
+
+
+  template<typename T>
+  TableExprNodeSetOptContSetOC<T>::TableExprNodeSetOptContSetOC
+  (const TableExprNodeSet& orig,
+   const std::vector<T>& starts,
+   const std::vector<T>& ends)
+    : TableExprNodeSetOptContSet<T> (orig, starts, ends)
+  {}
+
+  template<typename T>
+  void TableExprNodeSetOptContSetOC<T>::show (ostream& os, uInt indent) const
+  {
+    TableExprNodeSetOptContSet<T>::show (os, indent);
+    os << " as TableExprNodeSetOptContSetOC" << endl;
+  }
+  
+  template<typename T>
+  Int64 TableExprNodeSetOptContSetOC<T>::find (T value) const
+  {
+    // Use lambda function to use <= instead of <.
+    auto iter = std::upper_bound (this->itsEnds.begin(), this->itsEnds.end(), value,
+                                  [](T v1, T v2){return v1<=v2;});
+    if (iter != this->itsEnds.end()) {
+      size_t index = std::distance (this->itsEnds.begin(), iter);
+      if (value > this->itsStarts[index]) {
+        return index;
+      }
+    }
+    return -1;
+  }
+
+
+  template<typename T>
+  TableExprNodeSetOptContSetOO<T>::TableExprNodeSetOptContSetOO
+  (const TableExprNodeSet& orig,
+   const std::vector<T>& starts,
+   const std::vector<T>& ends)
+    : TableExprNodeSetOptContSet<T> (orig, starts, ends)
+  {}
+
+  template<typename T>
+  void TableExprNodeSetOptContSetOO<T>::show (ostream& os, uInt indent) const
+  {
+    TableExprNodeSetOptContSet<T>::show (os, indent);
+    os << " as TableExprNodeSetOptContSetOO" << endl;
+  }
+  
+  template<typename T>
+  Int64 TableExprNodeSetOptContSetOO<T>::find (T value) const
+  {
+    auto iter = std::upper_bound (this->itsEnds.begin(), this->itsEnds.end(), value);
+    if (iter != this->itsEnds.end()) {
+      size_t index = std::distance (this->itsEnds.begin(), iter);
+      if (value > this->itsStarts[index]) {
+        return index;
+      }
+    }
+    return -1;
+  }
+
+
+  // Instantiate for Double and String.
+  template class TableExprNodeSetOptUSet<Int64>;
```

Comment:
This differs from the comment.
Can you explain why only these types are instantiated?

Reply:
Explained.